pkg/util/ip_reservation.go

```diff
@@ -0,0 +1,190 @@
+	pendingIPRangesMutex sync.Mutex
+}
+
+// NewIPAllocator is the constructor to initialize the IPAllocator object
+// Argument pendingIPRanges map[string]bool is a set of  IP ranges currently reserved by service instances but pending reservation in the cloud instances
+func NewIPAllocator(pendingIPRanges map[string]bool) *IPAllocator {
+	return &IPAllocator{
+		pendingIPRanges: pendingIPRanges,
+	}
+}
+
+// holdIPRange adds a particular IP range in the pendingIPRanges set
+// Argument ipRange string is an IPV4 range which needs put in pendingIPRanges
+func (ipAllocator *IPAllocator) holdIPRange(ipRange string) {
+	ipAllocator.pendingIPRanges[ipRange] = true
+}
+
+// ReleaseIPRange releases the pending IPRange
+// Argument ipRange string is an IPV4 range which needs to be released
+func (ipAllocator *IPAllocator) ReleaseIPRange(ipRange string) {
+	ipAllocator.pendingIPRangesMutex.Lock()
+	defer ipAllocator.pendingIPRangesMutex.Unlock()
+	delete(ipAllocator.pendingIPRanges, ipRange)
+}
+
+// ValidateCIDR function validates whether a particular cidr is a valid IP range
+// Argument cidr string is is a CIDR range that needs to be validated
+func (ipAllocator *IPAllocator) ValidateCIDR(cidr string) bool {
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return false
+	}
+	// The cidr network id size needs to be maximum 29 bits
+	cidrSize, _ := ipnet.Mask.Size()
+	if cidrSize > ipRangeSize {
+		return false
+	}
+
+	ipRangeMask := net.CIDRMask(ipRangeSize, ipV4Bits)
+	// The IP specified in the reserved-ipv4-cidr must be the first IP in the CIDR block
+	return ip.String() == ip.Mask(ipRangeMask).String()
+}
+
+// GetUnreservedIPRange returns an unreserved /29 IP block.
+// cidr: Provided cidr address in which we need to look for an unreserved /29 IP range
+// cloudInstancesReservedIPRanges: All the used IP ranges in the cloud instances
+// All the used IP ranges in the service instances not updated in cloud instances is extracted from the pendingIPRanges list in the IPAllocator
+// Finally a final reservedIPRange list is created by merging these two lists
+// Potential error cases:
+// 1) No /29 IP range in the CIDR is unreserved
+// 2) Parsing the CIDR resulted in an error
+func (ipAllocator *IPAllocator) GetUnreservedIPRange(cidr string, cloudInstancesReservedIPRanges map[string]bool) (string, error) {
+	ip, ipnet, err := net.ParseCIDR(cidr)
+	if err != nil {
+		return "", err
+	}
+	var reservedIPRanges = make(map[string]bool)
+
+	// The final reserved list is obtained by combining the cloudInstancesReservedIPRanges list and the pendingIPRanges list in the ipAllocator
+	for cloudInstancesReservedIPRange := range cloudInstancesReservedIPRanges {
+		reservedIPRanges[cloudInstancesReservedIPRange] = true
+	}
+
+	// Calculating increment step size and mask for /29 IP range
+	incrementStep29IPRange := (byte)(math.Exp2(ipV4Bits - ipRangeSize))
```

Comment:
Can these calculations be consts since they never change?

Reply:
got const initializer byte(math.Exp2(ipV4Bits - ipRangeSize)) is not a constant error when tried so

Reply:
thought about it. but then all the functions can mutate it. can change it though if it is more efficient

Reply:
it will at least save on redundant calculations.